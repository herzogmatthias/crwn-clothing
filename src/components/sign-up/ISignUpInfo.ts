export interface ISignUpInfo {
  displayName: string;
  email: string;
  password: string;
  confirmPassword: string;
}
