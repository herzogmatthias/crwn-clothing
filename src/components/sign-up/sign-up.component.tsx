import * as React from "react";
import { ISignUpInfo } from "./ISignUpInfo";
import FormInput from "../form-input/form-input.component";
import { CustomButton } from "../custom-button/custom-button.component";
import { TitleContainer, SignUpContainer } from "./sign-up.styles";
import { signUpStart } from "../../redux/user/user.actions";
import { ConnectedProps, connect } from "react-redux";

type ISignUpProps = ConnectedProps<typeof connector>;

export function SignUp({ signUpStart }: ISignUpProps) {
  const [signUpInfo, setSignUpInfo] = React.useState<ISignUpInfo>({
    displayName: "",
    confirmPassword: "",
    email: "",
    password: "",
  });
  const { email, password, displayName, confirmPassword } = signUpInfo;

  const _handleSubmit = async (event: React.FormEvent) => {
    event.preventDefault();
    if (password !== confirmPassword && password != "") {
      alert("passwords don't match");
      return;
    }
    signUpStart(signUpInfo);
  };
  const _handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const { value, name } = event.target;
    setSignUpInfo((prevState) => ({ ...prevState, [name]: value }));
  };
  return (
    <SignUpContainer>
      <TitleContainer>I do not have an account</TitleContainer>
      <span>Sign up with your email and password</span>
      <form onSubmit={_handleSubmit}>
        <FormInput
          handleChange={_handleChange}
          type="text"
          name="displayName"
          value={displayName}
          label="Diplay Name"
          required
        ></FormInput>
        <FormInput
          handleChange={_handleChange}
          type="email"
          name="email"
          value={email}
          label="Email"
          required
        ></FormInput>
        <FormInput
          handleChange={_handleChange}
          type="password"
          name="password"
          value={password}
          label="Password"
          required
        ></FormInput>
        <FormInput
          handleChange={_handleChange}
          type="password"
          name="confirmPassword"
          value={confirmPassword}
          label="Confirm Password"
          required
        ></FormInput>
        <CustomButton type="submit">SIGN UP</CustomButton>
      </form>
    </SignUpContainer>
  );
}

const mapDispatchToProps = {
  signUpStart: signUpStart,
};

const connector = connect(null, mapDispatchToProps);

export default connect(null, mapDispatchToProps)(SignUp);
