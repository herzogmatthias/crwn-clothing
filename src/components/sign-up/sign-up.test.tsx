import React from "react";
import { mount, ReactWrapper } from "enzyme";
import { SignUp } from "./sign-up.component";
import { ISignUpInfo } from "./ISignUpInfo";
import { act } from "react-dom/test-utils";

describe("NON CONNECTED SIGN UP TEST", () => {
  let wrapper: ReactWrapper;
  const mockPw = "asdf123";
  const mockDisplayName = "asdf";
  const mockEmail = "asasdf@gmail.com";
  let mockSignUpStart = jest.fn<any, [ISignUpInfo]>();
  const mockProps = {
    signUpStart: mockSignUpStart,
  } as any;
  wrapper = mount(<SignUp {...mockProps}></SignUp>);

  test("render SignUp Component", () => {
    expect(wrapper).toMatchSnapshot();
  });
  test("set the password value", async () => {
    await global.actions(wrapper, () => {
      wrapper.find('FormInput[name="password"]').prop<any>("handleChange")({
        target: { value: mockPw, name: "password" },
      } as any);
    });

    expect(wrapper.find('FormInput[name="password"]').prop("value")).toEqual(
      mockPw
    );
  });
  test("submit should fail with two different passwords", async () => {
    await global.actions(wrapper, () =>
      wrapper.find('FormInput[name="password"]').prop<any>("handleChange")({
        target: { value: mockPw, name: "password" },
      } as any)
    );
    await global.actions(wrapper, () =>
      wrapper
        .find('FormInput[name="confirmPassword"]')
        .prop<any>("handleChange")({
        target: { value: mockPw + "asdf", name: "confirmPassword" },
      } as any)
    );
    const alert = jest.spyOn(window, "alert").mockImplementation(() => {});
    wrapper.find("form").simulate("submit");
    expect(alert).toHaveBeenCalled();
  });
  test("submit should pass with same passwords", async () => {
    const signUpInfo: ISignUpInfo = {
      email: mockEmail,
      displayName: mockDisplayName,
      confirmPassword: mockPw,
      password: mockPw,
    };
    await global.actions(wrapper, () =>
      wrapper.find('FormInput[name="password"]').prop<any>("handleChange")({
        target: { value: mockPw, name: "password" },
      } as any)
    );
    await global.actions(wrapper, () =>
      wrapper
        .find('FormInput[name="confirmPassword"]')
        .prop<any>("handleChange")({
        target: { value: mockPw, name: "confirmPassword" },
      } as any)
    );
    await global.actions(wrapper, () =>
      wrapper.find('FormInput[name="displayName"]').prop<any>("handleChange")({
        target: { value: mockDisplayName, name: "displayName" },
      } as any)
    );
    await global.actions(wrapper, () =>
      wrapper.find('FormInput[name="email"]').prop<any>("handleChange")({
        target: { value: mockEmail, name: "email" },
      } as any)
    );
    wrapper.find("form").simulate("submit");
    expect(mockSignUpStart).toHaveBeenCalledWith(signUpInfo);
  });
});
