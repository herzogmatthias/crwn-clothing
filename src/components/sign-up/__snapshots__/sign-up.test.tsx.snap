// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NON CONNECTED SIGN UP TEST render SignUp Component 1`] = `
<SignUp
  signUpStart={[MockFunction]}
>
  <styled.div>
    <div
      className="sc-AxhUy hnTNpP"
    >
      <styled.h2>
        <h2
          className="sc-AxgMl iLAkrD"
        >
          I do not have an account
        </h2>
      </styled.h2>
      <span>
        Sign up with your email and password
      </span>
      <form
        onSubmit={[Function]}
      >
        <FormInput
          handleChange={[Function]}
          label="Diplay Name"
          name="displayName"
          required={true}
          type="text"
          value=""
        >
          <styled.div>
            <div
              className="sc-AxjAm hendVB"
            >
              <styled.input
                name="displayName"
                onChange={[Function]}
                required={true}
                type="text"
                value=""
              >
                <input
                  className="sc-AxiKw eoQQwa"
                  name="displayName"
                  onChange={[Function]}
                  required={true}
                  type="text"
                  value=""
                />
              </styled.input>
              <styled.label
                length={0}
              >
                <label
                  className="sc-AxirZ hZnEDN"
                >
                  Diplay Name
                </label>
              </styled.label>
            </div>
          </styled.div>
        </FormInput>
        <FormInput
          handleChange={[Function]}
          label="Email"
          name="email"
          required={true}
          type="email"
          value=""
        >
          <styled.div>
            <div
              className="sc-AxjAm hendVB"
            >
              <styled.input
                name="email"
                onChange={[Function]}
                required={true}
                type="email"
                value=""
              >
                <input
                  className="sc-AxiKw eoQQwa"
                  name="email"
                  onChange={[Function]}
                  required={true}
                  type="email"
                  value=""
                />
              </styled.input>
              <styled.label
                length={0}
              >
                <label
                  className="sc-AxirZ hZnEDN"
                >
                  Email
                </label>
              </styled.label>
            </div>
          </styled.div>
        </FormInput>
        <FormInput
          handleChange={[Function]}
          label="Password"
          name="password"
          required={true}
          type="password"
          value=""
        >
          <styled.div>
            <div
              className="sc-AxjAm hendVB"
            >
              <styled.input
                name="password"
                onChange={[Function]}
                required={true}
                type="password"
                value=""
              >
                <input
                  className="sc-AxiKw kaLtRc"
                  name="password"
                  onChange={[Function]}
                  required={true}
                  type="password"
                  value=""
                />
              </styled.input>
              <styled.label
                length={0}
              >
                <label
                  className="sc-AxirZ hZnEDN"
                >
                  Password
                </label>
              </styled.label>
            </div>
          </styled.div>
        </FormInput>
        <FormInput
          handleChange={[Function]}
          label="Confirm Password"
          name="confirmPassword"
          required={true}
          type="password"
          value=""
        >
          <styled.div>
            <div
              className="sc-AxjAm hendVB"
            >
              <styled.input
                name="confirmPassword"
                onChange={[Function]}
                required={true}
                type="password"
                value=""
              >
                <input
                  className="sc-AxiKw kaLtRc"
                  name="confirmPassword"
                  onChange={[Function]}
                  required={true}
                  type="password"
                  value=""
                />
              </styled.input>
              <styled.label
                length={0}
              >
                <label
                  className="sc-AxirZ hZnEDN"
                >
                  Confirm Password
                </label>
              </styled.label>
            </div>
          </styled.div>
        </FormInput>
        <CustomButton
          inverted={false}
          isGoogleSignIn={false}
          type="submit"
        >
          <styled.button
            inverted={false}
            isGoogleSignIn={false}
            type="submit"
          >
            <button
              className="sc-AxhCb jMBtEZ"
              type="submit"
            >
              SIGN UP
            </button>
          </styled.button>
        </CustomButton>
      </form>
    </div>
  </styled.div>
</SignUp>
`;
